### DISCUSSION TITLE
Question : support for win2d in WinUI 3 Desktop?

### DISCUSSION THREAD

**Discussion by jtorjo** [microsoft/WindowsAppSDK#756] [Q&A]
Now that #105 has been closed, since .Net 5 will never come to UWP, **when can we expect win2d support on WinUI 3 Desktop?**

This is the only thing keeping me from porting my app *BACK* to Desktop. My app was originally WPF, but had to switch to UWP because I need win2d. 

This is an insane blocker for me, and I'm sure for a lot of other people - without win2d, I can't really develop for WinUI 3 Desktop.

**Comment by duncanmacmichael** [comment]
Hi @jtorjo, appreciate the question. Win2D for WinUI 3 is a work in progress but we just released a preview package recently along with Project Reunion 0.5. 

Link to WinUI 3 current preview docs: https://docs.microsoft.com/en-us/windows/apps/winui/winui3/
Source: https://github.com/Microsoft/Win2D/tree/reunion_master
Documentation: http://microsoft.github.io/Win2D/WinUI3/html/Introduction.htm

The doc page and source Readme both have a link to the NuGet package to get started. Hope this helps!

**Reply by jtorjo** [reply]
@duncanmacmichael Thanks!

A few more questions:
Q1: Can I expect win2d to be fully ported to WinUI3 when Reunion 1.0 gets here?
Q2: On the doc for WinUI3, it mentions, as limitations, that `MediaElement` and `MediaPlayerElement` don't work. There's no mention as to when they would be ported to winUI3, and the roadmap doesn't mention them.

A few things:
1. Unless I'm too tired, I haven't seen anything pointing to a timetable about win2d. Even in the https://docs.microsoft.com/en-us/windows/apps/winui/winui3/ -- there's no mention about win2d whatsoever.
2. I need this, lets say, desperately. I'm totally fine waiting until Oct when Reunion 1 gets released, I just want to know it will be ready then
3. win2d, as it's written in the docs: "some features such as CanvasAnimatedControl or CanvasBitmap have partial or no support." -- those are both the cornerstone of the library - without them, I can't even begin to fathom porting my app to Desktop

**Reply by duncanmacmichael** [reply]
No problem! I'll answer as best I can here - this is a second take, since my first answer earlier wasn't accurate. Apologies if there was any confusion or if you get multiple notifications about this. :)
 
A1: We are working on getting more of Win2D supported for WinUI 3, but there are some fundamental changes we have to work through. For example, WinUI 3 has different capabilities, desktop is different from UWP, .NET 5 hits different codepaths, etc. We're working on these issues as we encounter them, but we might need a little more time and thus can't commit to an exact date.
A2: The WinUI 3 roadmap is the source of truth here, so I'd recommend checking back with them for the latest news.

A few things - answers:

1. We want to support our Win2D developers, which is why we are doing this work to get Win2D supported for WinUI3.  But the space we’re working in is in development and it’s not clear yet how best to support the developers.  Do we keep Win2D the way it looks today or should we refactor?  There may be advantages to rethinking its implementation/exposure.  We were hoping to have an answer for that by now, but it’s taken longer to get Win2D working for WinUI3 than we expected due to what I mentioned above.
2. Do you need full functionality or partial functionality? If you let us know what you need it can help prioritize different features as we work on it.
3. Understood. The CanvasBitmap issue is affected by this problem, which is being worked on: https://github.com/microsoft/CsWinRT/issues/612 We're working on CanvasAnimatedControl but it's not as far along in the fix. Stay tuned!

**Reply by llothar** [reply]
I only want to say that Direct2D or Win2D development API is fine for me, no need for changes. It takes time to understand it and the pitfalls of the current implementation but once you know it's ok. My practical unsolvable problems have been more with DirectWrite (look at Apples TextKit2 just announced at this WWDC21).

What is missing for both APIs are rational documentation of usage, not the API documentation which is fine, but explainations about the implementation pitfalls (like the builtin clipping not helping anything for performance issues) and day to day usages.

**Reply by mdtauk** [reply]
> I only want to say that Direct2D or Win2D development API is fine for me, no need for changes. It takes time to understand it and the pitfalls of the current implementation but once you know it's ok. My practical unsolvable problems have been more with DirectWrite (look at Apples TextKit2 just announced at this WWDC21).
> 
> What is missing for both APIs are rational documentation of usage, not the API documentation which is fine, but explainations about the implementation pitfalls (like the builtin clipping not helping anything for performance issues) and day to day usages.

DirectWrite has started an open source version, so hopefully that will help innovation and documentation

DWrite Core https://docs.microsoft.com/en-gb/windows/apps/project-reunion/dwritecore

**Reply by jtorjo** [reply]
@duncanmacmichael  Sorry for the late reply, I've been meaning to do this a long time ago!

A1:  While I totally understand that, it's still not very reassuring.
A2: I'm checking the Win3UI roadmap, every now and then :)

1. I really don't think you should refactor just yet. The top priority should be porting to WinUI 3 Desktop. Then, at some point in time, you may generate a new version with breaking changes. 
2. I need close to full functionality. Here's what I DON"T need: `Microsoft.Graphics.Canvas.UI.Composition`, `Microsoft.Graphics.Canvas.Printing`. Also, at this time, I'm not using `CanvasSpriteBatch`. Other than that, I'm using pretty much every other namespace.
3. Yes, there does seem to be work on this one. #612 is fixed, but there's another bug (#844) related to this. So hopefully this will work within 1-2 months :)

**Comment by castorix** [comment]
I just tested Direct2D in C# (P/Invoke) and it works fine
I just had to use a Region on "Microsoft.UI.Content.ContentWindowSiteBridge" window to display the part of the main window where I draw with ID2D1HwndRenderTarget.BeginDraw and so on
A test with _CLSID_D2D1Emboss_ : 

![WinUI_Direct2D](https://user-images.githubusercontent.com/22345506/142710680-fe5a04f1-21cd-4aaf-ad09-dc3cf20089c9.jpg)

**Reply by jtorjo** [reply]
Just to be clear -- using Direct2D directly is NOT win2d.

**Comment by JohnCido** [comment]
Hi, I was browsing the discussion and just saw this thread. I have a very simple bar chart in my app that utilized Win2D. I installed the NuGet package `Microsoft.Graphics.Win2D` and it works fine for my very limited usage.

You can check out the project page [here](https://microsoft.github.io/Win2D/WinUI3/html/Introduction.htm#!).
And you can see the current features [here](https://microsoft.github.io/Win2D/WinUI3/html/Features.htm).

**Comment by castorix** [comment]
> Just to be clear -- using Direct2D directly is NOT win2d.

Win2D is mainly a wrapper to Direct2D (ID2D1Factory2 and so on)

**Reply by jtorjo** [reply]
Obviously. But I want win2d to be ported to WinUI3 Desktop. I could have used Direct2D with pinvoke several years ago, but I don't want that hustle.
